Improve template and error message and  documentation (#1116)
IBM/unitxt: docs/docs/adding_template.rst

        output_format='{translation}',
    ),

> The `instruction` attribute defines that part of the prompt that appears one, while the `input_format` defines the part of prompt that repeats for 

Author: ```suggestion
The `instruction` attribute defines that part of the prompt that appears once, while the `input_format` defines the part of prompt that repeats for 
```

each in-context learning demonstration and for the final instance.  The `output_format` defines how the reference answer is verbalized as string.   
The InputOutputTemplate assumes there is at most only a single reference (gold answer). 
If you pass a field value which is a list to the InputOutputTemplate, then it is verbalized as comma separated string. For example, ["happy","angry"]

> become the string reference "happy,angry", and it is expected that the model will return that string as the correct answer.

Author: ```suggestion
becomes the string reference "happy,angry", and it is expected that the model will return that string as the correct answer.
```

> The `instruction` attribute defines that part of the prompt that appears once, while the `input_format` defines the part of prompt that repeats for 
> each in-context learning demonstration and for the final instance.  The `output_format` defines how the reference answer is verbalized as string.   
> The InputOutputTemplate assumes there is at most only a single reference (gold answer). 
> If you pass a field value which is a list to the InputOutputTemplate, then it is verbalized as comma separated string. For example, ["happy","angry"]
> becomes the string reference "happy,angry", and it is expected that the model will return that string as the correct answer.

Reviewer: ```suggestion
The ``instruction`` attribute defines that part of the prompt that appears once, while the ``input_format`` defines the part of prompt that repeats for 
each in-context learning demonstration and for the final instance.  The ``output_format`` defines how the reference answer is verbalized as string.   
The InputOutputTemplate assumes there is at most only a single reference (gold answer). 
If you pass a field value which is a list to the InputOutputTemplate, then it is verbalized as comma separated string. For example, ["happy","angry"]
becomes the string reference "happy,angry", and it is expected that the model will return that string as the correct answer.
```

.. note::
    If you don't have references , just specify `output_format=""`.
    If you have multiple references, use the MultiReferenceTemplate (see below)
    The only fields that are mandatory are the `input_format` and `output_format`

Post Processors
---------------

The template also defines the post processing steps done on the output predictions of the model before they are passed to the :ref:`Metrics <metric>`.

> The post processors applied both to the textual prediction and to the references. 

Author: [](url)
```suggestion
The post processors applied both to the model prediction and to the references. 
```

> The template also defines the post processing steps done on the output predictions of the model before they are passed to the :ref:`Metrics <metric>`.
> The post processors applied both to the model prediction and to the references. 
> For example, we could use the `processors.lower_case` processor to lowercase both the model predictions and references,
> so the metric computation will ignore case. 

Reviewer: ```suggestion
The template also defines the post processing steps applied to the output predictions of the model before they are passed to the :ref:`Metrics <metric>`.
The post processors applied both to the model prediction and to the references. 
For example, we could use the ``processors.lower_case`` processor to lowercase both the model predictions and references,
so the metric computation will ignore case. When needed, It is possible to add post processors that applied only to the output of the model and not the references or vice versa. 
```

> so the metrics computation will ignore case. 

Author: ```suggestion
so the metric computation will ignore case. 
```

    template = InputOutputTemplate(
        instruction="In the following task, you translate a {text_type}.",
        input_format="Translate this {text_type} from {source_language} to {target_language}: {text}.",
        target_prefix="Translation: ",
        output_format='{translation}',
        postprocessors= [
                "processors.lower_case"
            ]

>     )

Reviewer: should be in code block

Author: Done.

The reason the post processors are set in the template, is because different templates prompt the model to generate answers in different formats. 

> For example, one template may prompt the model to answer `Yes`` or `No` while another 

Reviewer: ```suggestion
For example, one template may prompt the model to answer `Yes` or `No` while another 
```

---------------

The template also defines the post processing steps applied to the output predictions of the model before they are passed to the :ref:`Metrics <metric>`.
The post processors applied both to the model prediction and to the references. 
For example, we could use the ``processors.lower_case`` processor to lowercase both the model predictions and references,
so the metric computation will ignore case. When needed, It is possible to add post processors that applied only to the output of the model and not the references or vice versa. 

.. code-block:: python

    template = InputOutputTemplate(
        instruction="In the following task, you translate a {text_type}.",
        input_format="Translate this {text_type} from {source_language} to {target_language}: {text}.",
        target_prefix="Translation: ",
        output_format='{translation}',
        postprocessors= [
                "processors.lower_case"
            ]
    )

The reason the post processors are set in the template, is because different templates prompt the model to generate answers in different formats. 
For example, one template may prompt the model to answer ``Yes`` or ``No`` while another 

> template may prompt the model to answer ``True`` or ``False``. Both can use different post processors to convert them to standard model prediction of `0` or `1`

Author: ```suggestion
template may prompt the model to answer ``True`` or ``False``. Both can use different post processors to convert them to standard model prediction of `0` or `1`.
```
